### ISSUE TITLE
toIncrementalJson results in "Too much Recursion" for certain types of objects

### ISSUE THREAD

**Issue by Glidias** [NorthwoodsSoftware/GoJS#21]
Anyway around this? It seems the codebase doesn't use JSON.stringify to handle the JSON creation, because I created a custom "toJSON" method to prevent recursion, but the output for toIncrementalJSON (it uses the full output) differs from the one in JSON.stringify.

http://stackoverflow.com/a/3895474

**Comment by WalterNorthwoods** [comment]
So you actually have cyclical references in your data?  I do not believe that either **Model.toJson** or Model.toIncrementalJson** check for that.

For now, I do not have a good work-around for you.  Are you depending on maintaining those cyclical or shared references when loading the model back in?

**Comment by simonsarris** [comment]
If you do really think there is a bug, some data that we can experiment with would be very helpful.

**Comment by Glidias** [comment]
Hmm.. saw this on the documentation and i guess it's a limitation that i have to get around with, I'm trying to integrate GOJS with a graphing library which I'm using as my model data and my data format involves managing of shared model properties and such...

In Model documentation...

> Note that JSON and other textual data formats cannot faithfully store all JavaScript functions. toJson and Model.fromJson do not try to save and load functional property values. You should arrange that all such functions, including event handlers, are established by your app. toJson and Model.fromJson also cannot handle circular references; any sharing of references will be lost too. They also skip properties that are not enumerable, those whose names start with an underscore, and those whose values are undefined.

eg.

     GoJS Model GraphNode
           -> LocationPacket (unique per node) -> 
                   LocationDefinition (shared content between nodes), 
                   LocationDefinitionOverwrites (shared or unique content  among nodes),   
                   LocationState (unique per node).

Basically, I need a binding system that allows changing one LocationDefinition's details to affect the visuals of all related Nodes that are using that same definition reference. But the data-binding paradigm is purely 1 to 1...without the possibly of sharing references. I guess i have to run through the model manually to update those nodes with a certain definition `e.model.updateTargetBindings(o, "text");`.

**Comment by WalterNorthwoods** [comment]
I was going to suggest making the "back-pointer" properties be non-enumerable, or prefixing them with "_", but I see now that you have updated your comment to anticipate that possibility.

Yes, using getter/setter properties is another way to solve the problem.  It looks like you have a good grasp on the situation.  If you have any other questions, contact us through our forum, https://forum.nwoods.com/c/gojs, or by email.
